### DISCUSSION TITLE
Are all asserts from a rejected solver step supposed to be hidden?

### DISCUSSION THREAD

**Discussion by bilderbuchi** [OpenModelica/OpenModelica#10798] [Q&A] [answered]
I have a model here which gives me a warning (yellow text) about a negative sqrt argument. As far as I can tell (from final plots etc), this negative argument is not real and happens during a rejected step. With `LOG_SOLVER` on:

> call solver from 0.142157 to 0.15 (stepSize: 0.00784270294097295)
finished solver step 0.150638
call solver from 0.150638 to 0.16 (stepSize: 0.00936195052153388)
The following assertion has been violated at time 0.158271
Model error: Argument of sqrt(kappa * p * (1.0 - (p_amb / p) ^ ((-1.0 + kappa) / kappa)) / ((-1.0 + kappa) * rho_b)) was -4777.33 should be >= 0
finished solver step 0.152546
call solver from 0.152546 to 0.16 (stepSize: 0.00745378122716009)
finished solver step 0.156363

The time 0.158271 is beyond the finished solver step, so I conclude that this happened in a solver step that was rejected as too large. I would have assumed that all asserts from such would be hidden, as they are not really actionable for the user (#6195 seems related).
AFAICT from the log (also with `LOG_EVENTS`), no events are involved in those steps, although the solver later finds a state event at time=0.156817411991 (so *before* the assert's time, but after that step was already concluded).

Is it expected that the negative-sqrt-arg assert was shown, or is that a bug (maybe this category of asserts was overlooked)?

NB: I tried to reduce my model to an MWE I can share, but failed so far to meaningfully reduce it.

**Comment by casella** [comment, marked as answer]
@bilderbuchi ideally the answer is YES.

However, it's not clear how to do that. The master solver attempts a step, by calling generated code that does something verboten, e.g. sqrt(neg). This triggers a warning and throws an exception, which is caught by the solver that discards the step and retries with a shorter one.

The problem is, it's not the solver that issues the warning, it's the low-level generated code that does it. And, unfortunately, there's no way it can be aware of how this error will be interpreted.

In principle, we should not print out warnings immediately, but rather put them in a bottle and pass them up to the caller, which will decide whether to post the messag in the bottle or discard it. This would require a massive redesign of the generated code and runtime.

BTW, also other tools (e.g. Dymola) do the same, I guess for the same reason...

**Reply by bilderbuchi** [reply]
Thanks for the context! Makes sense, I failed to consider how user-made asserts are different from internal ones, that makes sense and I can see how that can be hard to restructure if the code has not been set up with that in mind. 
This was mainly weird noise that turned up in a working model, and I tried to find out if there's a modelling error behind it, no big deal.

**Reply by casella** [reply]
There's actually something even worse, it's the "Matrix singular!!!" (in red, with three exclamation marks) message, which I find particularly irritating because it doesn't tell me _which_ matrix is singular, so it's totally useless, except for making you feel uneasy about your simulation.

But if you think about it, this message comes from a poor quasi-Newton algorithm, which just sees a vector of residuals and a Jacobian. How could it know what this matrix means and where does it come from?

**Comment by casella** [comment]
#6195 is different. It applies additional constraints (assertions) which are written by the modeller that knows what are the boundaries of validity of the model. There's no sense checking them during iterations: if they are violated, the result will probably not be reliable, but that's ok during iterations, as long as they eventually converge to a solution which does not trigger the warning.

Sqrt(neg), log(neg) are in a different league. You can't ignore therm and carry on with the iterations, as they will produce NaNs that will pollute the solution like a cancer. So, they just need to be discarded.

**Reply by casella** [reply]
> #6195 is different. It applies additional constraints (assertions) which are written by the modeller that knows what are the boundaries of validity of the model. There's no sense checking them during iterations: if they are violated, the result will probably not be reliable, but that's ok during iterations, as long as they eventually converge to a solution which does not trigger the warning.

In fact, collocation methods for optimal control are based on this concept: the model equations are fulfilled only at convergence, so during iterations basic principles like mass or energy conservation are happily violated. This is OK as long as you eventually converge.

Sqrt(neg) and log(neg) are in a different league. You can't ignore therm and carry on with the iterations, as they will produce NaNs that will pollute the solution like a cancer. So, they just need to be discarded.
